# Introduction To Pandas
## The Basics Of DataFrames

#### Importing pandas

In [91]:
import pandas as pd

#### Reading data into a DataFrame

In [92]:
red_wine_quality = pd.read_csv("data/winequality-red.csv", delimiter=';')

#### Getting a sense of the data

##### How many rows and columns are in the DataFrame?

In [93]:
red_wine_quality.shape

(1599, 12)

- How many rows and columns are in the DataFrame?
> The DataFrame includes `1599` rows and `12` columns.

In [94]:
for col in red_wine_quality.columns:
    print(col)

fixed acidity
volatile acidity
citric acid
residual sugar
chlorides
free sulfur dioxide
total sulfur dioxide
density
pH
sulphates
alcohol
quality


##### What data type is in each column?

| Column Name          | Data Types |
|----------------------|------------|
| fixed acidity        | float64    |
| volatile acidity     | float64    |
| citric acid          | float64    |
| residual sugar       | float64    |
| chlorides            | float64    |
| free sulfur dioxide  | float64    |
| total sulfur dioxide | float64    |
| density              | float64    |
| pH                   | float64    |
| sulphates            | float64    |
| alcohol              | float64    |
| quality              | int64      |

##### Are all of the variables continuous, or are any categorical?
> All of the variables are contiuous.

##### How many non-null values are in each column?

In [95]:
red_wine_quality.isna().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

##### What are the min, mean, max, median for all numeric columns?

In [96]:
red_wine_quality.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

In [97]:
red_wine_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null int64
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


## Practice with Grabbing Data

##### Grab the first 10 rows of the `chlorides` column. 

In [98]:
red_wine_quality[['chlorides']][0:10]

chlorides
0      0.076
1      0.098
2      0.092
3      0.075
4      0.076
5      0.075
6      0.069
7      0.065
8      0.073
9      0.071

##### Grab the last 10 rows of the `chlorides` column. 

In [99]:
red_wine_quality[['chlorides']][-10:]

chlorides
1589      0.073
1590      0.077
1591      0.089
1592      0.076
1593      0.068
1594      0.090
1595      0.062
1596      0.076
1597      0.075
1598      0.067

##### Grab indices 264-282 of the `chlorides` **and** `density` columns. 

In [100]:
red_wine_quality[['chlorides', 'density']][264:282]

chlorides  density
264      0.064  0.99990
265      0.071  0.99680
266      0.096  1.00025
267      0.078  0.99730
268      0.077  0.99870
269      0.104  0.99960
270      0.087  0.99650
271      0.104  0.99960
272      0.071  0.99935
273      0.076  0.99735
274      0.088  0.99915
275      0.087  0.99650
276      0.077  0.99870
277      0.104  0.99960
278      0.073  0.99760
279      0.087  0.99910
280      0.071  0.99860
281      0.358  0.99720

##### Grab all rows where the `chlorides` value is less than 0.10.

In [101]:
is_lower = red_wine_quality['chlorides'] < .1
filtered = red_wine_quality[is_lower]
filtered[['chlorides']]

chlorides
0         0.076
1         0.098
2         0.092
3         0.075
4         0.076
5         0.075
6         0.069
7         0.065
8         0.073
9         0.071
10        0.097
11        0.071
12        0.089
16        0.092
18        0.086
20        0.077
21        0.082
23        0.084
24        0.085
25        0.080
26        0.080
28        0.080
29        0.082
30        0.089
32        0.083
33        0.073
35        0.086
36        0.086
37        0.066
39        0.074
...         ...
1567      0.076
1568      0.077
1569      0.056
1571      0.038
1572      0.069
1573      0.075
1574      0.074
1575      0.060
1576      0.081
1577      0.076
1579      0.053
1580      0.068
1581      0.053
1582      0.053
1583      0.074
1584      0.061
1585      0.066
1586      0.065
1587      0.066
1588      0.068
1589      0.073
1590      0.077
1591      0.089
1592      0.076
1593      0.068
1594      0.090
1595      0.062
1596      0.076
1597      0.075
1598      0.067

[1363 rows x 1 columns]

##### Now grab all the rows where the `chlorides` value is greater than the column's mean (try **not** to use a hard-coded value for the mean, but instead a method).

In [102]:
red_wine_quality['chlorides'].mean()

0.08746654158849279

In [103]:
def row_value_is_greater_than(df, column_name, value):
    condition = df[column_name] > value
    filtered = df[condition]
    return filtered[[column_name]]

In [104]:
def is_bigger_than_mean(df, column_name):
    return row_value_is_greater_than(df, column_name, df[column_name].mean())

In [105]:
is_bigger_as_mean(red_wine_quality, 'chlorides')

chlorides
1         0.098
2         0.092
10        0.097
12        0.089
13        0.114
14        0.176
15        0.170
16        0.092
17        0.368
19        0.341
22        0.106
27        0.106
30        0.089
31        0.105
34        0.103
38        0.172
41        0.088
42        0.332
46        0.114
47        0.113
54        0.110
57        0.111
61        0.115
71        0.088
74        0.094
75        0.093
76        0.093
77        0.104
81        0.464
83        0.401
...         ...
1391      0.094
1395      0.092
1396      0.090
1413      0.104
1423      0.123
1434      0.214
1435      0.214
1436      0.169
1441      0.094
1445      0.094
1451      0.104
1474      0.205
1476      0.205
1479      0.104
1481      0.104
1482      0.114
1493      0.089
1496      0.089
1504      0.106
1507      0.106
1518      0.114
1519      0.106
1528      0.096
1548      0.099
1552      0.103
1558      0.235
1570      0.230
1578      0.118
1591      0.089
1594      0.090

[474 rows x 1 columns]

##### Grab all those rows where the `pH` is greater than 3.0 and less than 3.5.

In [106]:
column_name = 'pH'
left_value = 3.0
right_value = 3.5

red_wine_quality['pH'].between(3, 3.5, inclusive=True)
red_wine_quality

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
5               7.4             0.660         0.00             1.8      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
8               7.8             0.580         0.02             2.0      0.073   
9               7.5             0.500         0.36             6.1      0.071   
10              6.7             0.580         0.08             1.8      0.097   
11              7.5             0.500         0.36             6.1      0.071   
12              5.6             0.615         0.00             1.6      0.089   
13              7.8             0.610         0.29             1.6      0.114   
14              8.9             0.620         0.18             3.8      0.176   
15              8.9             0.620         0.19             3.9      0.170   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
18              7.4             0.590         0.08             4.4      0.086   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
21              7.6             0.390         0.31             2.3      0.082   
22              7.9             0.430         0.21             1.6      0.106   
23              8.5             0.490         0.11             2.3      0.084   
24              6.9             0.400         0.14             2.4      0.085   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
29              7.8             0.645         0.00             2.0      0.082   
...             ...               ...          ...             ...        ...   
1569            6.2             0.510         0.14             1.9      0.056   
1570            6.4             0.360         0.53             2.2      0.230   
1571            6.4             0.380         0.14             2.2      0.038   
1572            7.3             0.690         0.32             2.2      0.069   
1573            6.0             0.580         0.20             2.4      0.075   
1574            5.6             0.310         0.78            13.9      0.074   
1575            7.5             0.520         0.40             2.2      0.060   
1576            8.0             0.300         0.63             1.6      0.081   
1577            6.2             0.700         0.15             5.1      0.076   
1578            6.8             0.670         0.15             1.8      0.118   
1579            6.2             0.560         0.09             1.7      0.053   
1580            7.4             0.350         0.33             2.4      0.068   
1581            6.2             0.560         0.09             1.7      0.053   
1582            6.1             0.715         0.10             2.6      0.053   
1583            6.2             0.460         0.29             2.1      0.074   
1584            6.7             0.320         0.44             2.4      0.061   
1585            7.2             0.390         0.44             2.6      0.066   
1586            7.5             0.310

In [107]:
column_name = 'pH'
left_value = 3.0
right_value = 3.5

is_greater_than = 3.0 < red_wine_quality['pH']
filtered = red_wine_quality[is_greater_than]
is_less_than = red_wine_quality['pH'] < 3.5
filtered = red_wine_quality[is_less_than]
filtered

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
8               7.8             0.580         0.02             2.0      0.073   
9               7.5             0.500         0.36             6.1      0.071   
10              6.7             0.580         0.08             1.8      0.097   
11              7.5             0.500         0.36             6.1      0.071   
13              7.8             0.610         0.29             1.6      0.114   
14              8.9             0.620         0.18             3.8      0.176   
15              8.9             0.620         0.19             3.9      0.170   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
18              7.4             0.590         0.08             4.4      0.086   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
22              7.9             0.430         0.21             1.6      0.106   
23              8.5             0.490         0.11             2.3      0.084   
24              6.9             0.400         0.14             2.4      0.085   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
29              7.8             0.645         0.00             2.0      0.082   
30              6.7             0.675         0.07             2.4      0.089   
31              6.9             0.685         0.00             2.5      0.105   
32              8.3             0.655         0.12             2.3      0.083   
33              6.9             0.605         0.12            10.7      0.073   
34              5.2             0.320         0.25             1.8      0.103   
...             ...               ...          ...             ...        ...   
1560            7.8             0.600         0.26             2.0      0.080   
1561            7.8             0.600         0.26             2.0      0.080   
1562            7.2             0.695         0.13             2.0      0.076   
1563            7.2             0.695         0.13             2.0      0.076   
1564            7.2             0.695         0.13             2.0      0.076   
1565            6.7             0.670         0.02             1.9      0.061   
1566            6.7             0.160         0.64             2.1      0.059   
1567            7.2             0.695         0.13             2.0      0.076   
1568            7.0             0.560         0.13             1.6      0.077   
1569            6.2             0.510         0.14             1.9      0.056   
1570            6.4             0.360         0.53             2.2      0.230   
1571            6.4             0.380         0.14             2.2      0.038   
1572            7.3             0.690         0.32             2.2      0.069   
1574            5.6             0.310         0.78            13.9      0.074   
1575            7.5             0.520         0.40             2.2      0.060   
1576            8.0             0.300         0.63             1.6      0.081   
1578            6.8             0.670         0.15             1.8      0.118   
1580            7.4             0.350

##### Further filter the results from 6 to grab only those rows that have a `residual sugar` less than 2.0. 

In [108]:
has_less_sugar = filtered['residual sugar'] < 2.0
filtered = filtered[has_less_sugar]
filtered

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
3              11.2             0.280         0.56             1.9      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
10              6.7             0.580         0.08             1.8      0.097   
13              7.8             0.610         0.29             1.6      0.114   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
22              7.9             0.430         0.21             1.6      0.106   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
34              5.2             0.320         0.25             1.8      0.103   
44              6.8             0.670         0.02             1.8      0.050   
47              8.7             0.290         0.52             1.6      0.113   
48              6.4             0.400         0.23             1.6      0.066   
49              5.6             0.310         0.37             1.4      0.074   
50              8.8             0.660         0.26             1.7      0.074   
61              7.7             0.690         0.49             1.8      0.115   
62              7.5             0.520         0.16             1.9      0.085   
66              7.5             0.520         0.11             1.5      0.079   
67              6.6             0.705         0.07             1.6      0.076   
69              8.0             0.705         0.05             1.9      0.074   
70              7.7             0.630         0.08             1.9      0.076   
72              7.7             0.690         0.22             1.9      0.084   
79              8.3             0.625         0.20             1.5      0.080   
80              6.2             0.450         0.20             1.6      0.069   
81              7.8             0.430         0.70             1.9      0.464   
...             ...               ...          ...             ...        ...   
1458            8.4             0.290         0.40             1.7      0.067   
1459            7.9             0.200         0.35             1.7      0.054   
1464            6.8             0.590         0.10             1.7      0.063   
1465            6.8             0.590         0.10             1.7      0.063   
1470           10.0             0.690         0.11             1.4      0.084   
1473            6.1             0.600         0.08             1.8      0.071   
1483            8.1             0.330         0.44             1.5      0.042   
1490            7.1             0.220         0.49             1.8      0.039   
1493            7.7             0.540         0.26             1.9      0.089   
1494            6.4             0.310         0.09             1.4      0.066   
1495            7.0             0.430         0.02             1.9      0.080   
1496            7.7             0.540         0.26             1.9      0.089   
1500            7.5             0.725         0.04             1.5      0.076   
1509            7.9             0.180         0.40             1.8      0.062   
1513            6.4             0.560         0.15             1.8      0.078   
1521            6.9             0.480         0.20             1.9      0.082   
1527            7.1             0.530         0.07             1.7      0.071   
1533            8.0             0.390